Detectors/TPC/base/include/TPCBase/CDBInterface.h

```diff
@@ -31,12 +34,52 @@ class ParameterElectronics;
 class ParameterGas;
 class ParameterGEM;
 
+/// Calibration and parameter types for CCDB
+enum class CDBType {
+  CalPedestal,        ///< Pedestal calibration
+  CalNoise,           ///< Noise calibration
+  CalPulser,          ///< Pulser calibration
+  CalPadGainFull,     ///< Full pad gain calibration
+  CalPadGainResidual, ///< ResidualpPad gain calibration (e.g. from tracks)
+                      ///
+  ParDetector,        ///< Parameter for Detector
+  ParElectronics,     ///< Parameter for Electronics
+  ParGas,             ///< Parameter for Gas
+  ParGEM,             ///< Parameter for GEM
+};
+
+/// Upload intervention type
+enum class CDBIntervention {
+  Manual,    ///< Upload from manual intervention
+  Automatic, ///< Automatic upload
+};
+
+/// Storage name in CCDB for each calibration and parameter type
+const std::unordered_map<CDBType, std::string> CDBTypeMap{
+  {CDBType::CalPedestal, "TPC/Calib/Pedestal"},
+  {CDBType::CalNoise, "TPC/Calib/Noise"},
+  {CDBType::CalPulser, "TPC/Calib/Pulser"},
+  {CDBType::CalPadGainFull, "TPC/Calib/PadGainFull"},
+  {CDBType::CalPadGainResidual, "TPC/Calib/PadGainResidual"},
+  //
+  {CDBType::ParDetector, "TPC/Parameter/Detector"},
+  {CDBType::ParElectronics, "TPC/Parameter/Electronics"},
+  {CDBType::ParGas, "TPC/Parameter/Gas"},
+  {CDBType::ParGEM, "TPC/Parameter/GEM"},
+};
+
+/// Poor enum reflection ...
+const std::unordered_map<CDBIntervention, std::string> CDBInterventionMap{
+  {CDBIntervention::Manual, "Manual"},
+  {CDBIntervention::Automatic, "Automatic"},
+};
+
 /// \class CDBInterface
 /// The class provides a simple interface to the CDB for the TPC specific
 /// objects. It will not take ownership of the objects, but will leave this
 /// to the CDB itself.
```

Comment:
Actually, in opposite to aliroot, the ccdb itself does not own the object. So, to avoid lot of memory leaks, with the current implementation detectors must take the ownership of the objects they retrieve and take care of their cleanup when updating them for a new timestamp. 
But this may introduce an extra mess (like we had in aliroot) once we have different ways of accessing the ccdb: directly by detector, by the framework as a service etc. 
Another caveat is that for every new TF time stamp the object will be loaded again, although for most of objects the validity range will be >> than the TF.

What about making the ``BasicCCDBManager`` to own the objects it retrieves, together with its metadata, something like the code below (the caching may be done optional, if an update of the CCDB during the lifespan of the object being in use is expected):
````
using MetaData = std::map<std::string, std::string>;
using CachedObject = std::pair<void*, MetaData>;
//..
std::unordered_map<std::string, CachedObject> mCache;
//..
public:
template <typename T>
T* get(std::string const& path)
{
   auto entry = getEntry(path);
   if (!isValidEntry(entry)) {
      delete reinterpret_cast<T*>(entry.first);
      entry.first = mCCDBAccessor.retrieveFromTFileAny<T>(path, entry.second, mTimestamp);
   }
   return reinterpret_cast<T*>(entry.first);
}

private:
 bool isValidEnry(const CachedObject& obj) {
   return obj.first && <check in obj.second metadata if is it good for mTimestamp>;  // will require adding validity info to MetaData
 }
}

@sawenzel @costing @chiarazampolli  what is your opinion?
````

Reply:
I think we need a different kind of caching, where the entry path is mapped to a unique identifier (CCDB's uuid). The code should still ask the server whether the entry it has is still valid for the new TF, presenting the cached uuid to the request. In case there is nothing newer than that the server will reply with the HTTP 304 (Not modified) code, and no content. Only then the client should be free to reuse the cached object. See slide 11 of [this](https://docs.google.com/presentation/d/1RMIzqHL1JnDhwmqGj_yTmxqjNb54hNoJwvFIgtldR6g/edit#slide=id.g39127e45c1_0_45) presentation.

Otherwise, yes, the ownership should stay with the framework. Last time we discussed this with @ktf the best approximation for this would be another device that would handle the ccdb interaction and cache the objects on behalf of everybody needing calibration data. I'll let Giulio comment on this.

Reply:
@costing Thanks! The query to CCDB will not necessarily go via DPL framework, validity checks might be needed also from macros or executables. So, the natural would be to give the ownership to e.g. BasicCCDBManager and make the DPL to query via it. 
As for the method to check the validity: the snippet above would still remain except that the ``isValidEnry`` would need to query the server rather than the metadata.
I guess this uuid should be generated on the server side and shipped with retrieved object (in the headers?)? Can you add such functionality to CcdbApi?

Reply:
It already is there, the corresponding metadata key is called "ETag", this is the unique identifier assigned by ccdb to each object (slides 7 and 10 of the above presentation).

I think a bit more than the validity check needs to change. When asking the server side, the answer can be either "Go ahead, use what you have, no data", or "Here is the new content". All happens in a single step, avoiding making two calls for separately checking the validity and fetching the new content.

Reply:
Thanks, will check.

Reply:
Thanks for the feedback and discussion!